File: app/src/main/java/com/anadi/weatherinfo/AddLocationActivity.java
package com.anadi.weatherinfo;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.Toast;

import java.util.ArrayList;

import timber.log.Timber;

public class AddLocationActivity extends AppCompatActivity implements Spinner.OnItemSelectedListener {

    Spinner countrySpinner;
    Spinner citySpinner;
    Button addLocationButton;

    private String selectedCountry;
    private String selectedCity;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_location);

        ArrayList<String> countries = new ArrayList<String>();

        countries = Location.getCountryNamesArray();


        countrySpinner = findViewById(R.id.country_spinner);
        citySpinner = findViewById(R.id.city_spinner);
        addLocationButton = findViewById(R.id.add_location_button);

        countrySpinner.setOnItemSelectedListener(this);
        citySpinner.setOnItemSelectedListener(this);

        countrySpinner.setAdapter(new ArrayAdapter<String>(this, android.R.layout.simple_spinner_dropdown_item, countries ));
    }


    public void addLocation(View view) {

        if (TextUtils.isEmpty(selectedCity) || TextUtils.isEmpty(selectedCountry) ||
                selectedCity.equalsIgnoreCase("Select Item") ||
                selectedCountry.equalsIgnoreCase("Select Item")) {
            Timber.d( "selectedCity: " + selectedCity +
                                           "selectedCountry: " + selectedCountry);
            return;
        }

        if (!CitiesCash.add(selectedCity, selectedCountry)) {
            Toast.makeText(getApplicationContext(),
                    "Can`t load info for:" + selectedCity + " (selectedCountry).",
                    Toast.LENGTH_LONG).show();
        }

        finish();
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {

        if (parent.getId() == R.id.country_spinner) {
            selectedCountry = countrySpinner.getItemAtPosition(position).toString();

            if(!selectedCountry.equalsIgnoreCase("Select Item"))
                setCitySpinner();
        }

        if (parent.getId() == R.id.city_spinner)
            selectedCity = citySpinner.getItemAtPosition(position).toString();
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }

    private void setCitySpinner() {
        ArrayList<String> cities = Location.getCityNames(selectedCountry);


        if (null == cities) {

            Timber.d( "EMPTY cities LIST");
            return;
        }
        else
            Timber.d( cities.toString());

        citySpinner.setAdapter(new ArrayAdapter<String>(this, android.R.layout.simple_spinner_dropdown_item, cities ));
    }
}

File: app/src/main/java/com/anadi/weatherinfo/ui/addlocation/AddLocationActivity.kt
package com.anadi.weatherinfo.ui.addlocation

import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.*
import android.widget.AdapterView.OnItemSelectedListener
import androidx.appcompat.app.AppCompatActivity
import by.kirich1409.viewbindingdelegate.viewBinding
import com.anadi.weatherinfo.R
import com.anadi.weatherinfo.WeatherApplication
import com.anadi.weatherinfo.databinding.ActivityAddLocationBinding
import javax.inject.Inject

class AddLocationActivity : AppCompatActivity(R.layout.activity_add_location), OnItemSelectedListener, AddLocationContract.View {

    private val binding by viewBinding(ActivityAddLocationBinding::bind, R.id.layout)

    @Inject
    lateinit var presenter: AddLocationContract.Presenter

    private var selectedCountry = ""
    private var selectedCity = ""

    override fun loading() {
        binding.progress.visibility = View.VISIBLE
    }

    override fun onError(resId: Int) {
        binding.progress.visibility = View.GONE
        Toast.makeText(applicationContext, getText(resId), Toast.LENGTH_LONG).show()
    }

    override fun onAddedSuccess() {
        binding.progress.visibility = View.GONE
        finish()
    }

    override fun updateCityList(cities: List<String>) {
        binding.citySpinner.adapter = ArrayAdapter(this, R.layout.spinner_dropdown_item, cities)
    }

    fun addLocation(view: View?) {
        presenter.addLocation(selectedCity, selectedCountry)
    }

    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        if (parent.id == R.id.country_spinner) {
            selectedCountry = binding.countrySpinner.getItemAtPosition(position).toString()
            if (!TextUtils.isEmpty(selectedCountry) && !selectedCountry.equals("Select Item", ignoreCase = true)) {
                presenter.onCountrySelected(selectedCountry)
            }
        }
        if (parent.id == R.id.city_spinner) {
            selectedCity = binding.citySpinner.getItemAtPosition(position).toString()
        }
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {}

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        WeatherApplication.graph.inject(this)
        presenter.setView(this)

        binding.countrySpinner.onItemSelectedListener = this
        binding.citySpinner.onItemSelectedListener = this
        binding.countrySpinner.adapter = ArrayAdapter(this, R.layout.spinner_dropdown_item,
                presenter.countryNames)
    }
}